### PR TITLE
Features/federation database meta data test

### PR DESCRIPTION
**Pull request by aoyvx**
Fixes #[17375](https://github.com/apache/shardingsphere/issues/17375).

Changes proposed in this pull request:
- add FederationDatabaseMetaDataTest.java
- fix Use the nonexistent name to get `SchemaMetadata` will throw an `NullPointerException`

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import org.apache.groovy.util.Maps;
+import org.apache.shardingsphere.infra.metadata.database.schema.decorator.model.ShardingSphereSchema;
+import org.apache.shardingsphere.infra.metadata.database.schema.decorator.model.ShardingSphereTable;
+import org.junit.Test;
+
+public class FederationDatabaseMetaDataTest {
+    
+    @Test
+    public void assertPutSchemaMetadata() {
+        final FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo",
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:39]: @aoyvx Please remove final modifier in method domain.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                Collections.emptyMap());
+        final FederationSchemaMetaData schemaMetaData = mock(FederationSchemaMetaData.class);
+        federationDatabaseMetaData.putSchemaMetadata("foo_db", schemaMetaData);
+        assertThat(federationDatabaseMetaData.getSchemas().get("foo_db"), is(schemaMetaData));
+    }
+    
+    @Test
+    public void assertRemoveSchemaMetadata() {
+        final Map<String, ShardingSphereSchema> map = Maps.of(
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:48]: @aoyvx Please remove final modifier in method domain.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by aoyvx** [comment]
@strongduanmu  done

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    @Test
+    public void assertRemoveTableMetadata() {
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", Collections.emptyMap());
+        ShardingSphereTable fooTable = new ShardingSphereTable("foo_table", Collections.emptyList(),
+                Collections.emptyList(), Collections.emptyList());
+        federationDatabaseMetaData.putTable("foo_db", fooTable);
+        federationDatabaseMetaData.removeTableMetadata("foo_db", "foo_table");
+        assertNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").map(s -> s.getTables().get("foo_table")).orElse(null));
+    }
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:93]: Please remove this useless black line.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo",
+                Collections.emptyMap());
+        FederationSchemaMetaData schemaMetaData = mock(FederationSchemaMetaData.class);
+        federationDatabaseMetaData.putSchemaMetadata("foo_db", schemaMetaData);
+        assertThat(federationDatabaseMetaData.getSchemas().get("foo_db"), is(schemaMetaData));
+    }
+    
+    @Test
+    public void assertRemoveSchemaMetadata() {
+        Map<String, ShardingSphereSchema> map = Maps.of(
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:48]: Please replace `Maps.of()` with `new HashMap()`;

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        assertTrue(federationDatabaseMetaData.getSchemas().containsKey("foo_db_1"));
+        federationDatabaseMetaData.removeSchemaMetadata("foo_db_1");
+        assertFalse(federationDatabaseMetaData.getSchemas().containsKey("foo_db_1"));
+        assertTrue(federationDatabaseMetaData.getSchemas().containsKey("foo_db_2"));
+    }
+    
+    @Test
+    public void assertGetSchemaMetadata() {
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
+        Map<String, ShardingSphereSchema> map = Maps.of("foo_db", sphereSchema);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:61]: Please replace `Maps.of()` with `new HashMap()`;

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", map);
+        assertNotNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").orElse(null));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:63]: Do you think use assertTrue is better?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    
+    @Test
+    public void assertGetSchemaMetadataByNonexistentKey() {
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
+        Map<String, ShardingSphereSchema> map = Maps.of("foo_db", sphereSchema);
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", map);
+        federationDatabaseMetaData.getSchemaMetadata("foo_db_2");
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:71]: Please add assert for this case.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
+        Map<String, ShardingSphereSchema> map = Maps.of("foo_db", sphereSchema);
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", map);
+        assertNotNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").orElse(null));
+    }
+    
+    @Test
+    public void assertGetSchemaMetadataByNonexistentKey() {
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
+        Map<String, ShardingSphereSchema> map = Maps.of("foo_db", sphereSchema);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:69]: Please replace `Maps.of()` with `new HashMap()`;

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    public void assertGetSchemaMetadata() {
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
+        Map<String, ShardingSphereSchema> map = Maps.of("foo_db", sphereSchema);
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", map);
+        assertNotNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").orElse(null));
+    }
+    
+    @Test
+    public void assertGetSchemaMetadataByNonexistentKey() {
+        ShardingSphereSchema sphereSchema = mock(ShardingSphereSchema.class);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:68]: Rename sphereSchema to schema may be better.

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        federationDatabaseMetaData.getSchemaMetadata("foo_db_2");
+    }
+    
+    @Test
+    public void assertPutTable() {
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", Collections.emptyMap());
+        ShardingSphereTable fooTable = new ShardingSphereTable("foo_table", Collections.emptyList(),
+                Collections.emptyList(), Collections.emptyList());
+        federationDatabaseMetaData.putTable("foo_db", fooTable);
+        assertNotNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").orElse(null));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:80]: Do you think use assertTrue is better?

File: shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    
+    @Test
+    public void assertRemoveTableMetadata() {
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData("foo", Collections.emptyMap());
+        ShardingSphereTable fooTable = new ShardingSphereTable("foo_table", Collections.emptyList(),
+                Collections.emptyList(), Collections.emptyList());
+        federationDatabaseMetaData.putTable("foo_db", fooTable);
+        federationDatabaseMetaData.removeTableMetadata("foo_db", "foo_table");
+        assertNull(federationDatabaseMetaData.getSchemaMetadata("foo_db").map(s -> s.getTables().get("foo_table")).orElse(null));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/metadata/FederationDatabaseMetaDataTest.java:91]: Do you think use assertTrue is better?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by aoyvx** [comment]
@strongduanmu Thank you for your suggestion, I have done it here.

**Review by strongduanmu — APPROVED** [review]

**Comment by strongduanmu** [comment]
@aoyvx Thank you for your contribution.

**Pull request merged**